lib/esbonio/esbonio/lsp/__init__.py

```diff
@@ -179,11 +185,32 @@ def on_code_action(ls: RstLanguageServer, params: CodeActionParams):
 
         return actions
 
+    @server.feature(HOVER)
+    def on_hover(ls: RstLanguageServer, params: HoverParams):
+        uri = params.text_document.uri
+        doc = ls.workspace.get_document(uri)
+        pos = params.position
+        context = HoverContext(doc=doc, position=pos)
+
+        hover_values = []
+        for feature in ls._features.values():
```

Comment:
It might be worth copying parts of `on_completion` where it only calls `hover` on a feature if the line matches one of the patterns declared on it.

e.g. `hover` should only be called on the roles feature if the line actually contains a role.